File: 1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, vector<int>> indices;
        vector<int> ans;
        for (int i = 0; i < nums.size(); ++i) {
            indices[nums[i]].push_back(i);
        }
        for (auto i : indices) {
            if (indices.count(target - i.first)) {
                ans.push_back(indices[i.first][0]);
                if (target - i.first == i.first)
                    ans.push_back(indices[i.first][1]);
                else
                    ans.push_back(indices[target - i.first][0]);
                break;
            }
        }
        return ans;
    }
};
// @lc code=end

File: 1.Two-Sum.ts
function twoSum(nums: number[], target: number): number[] {
  let result: number[] = [];
  let index: Map<number, number[]> = new Map();
  for (let i = 0; i < nums.length; ++i) {
    if (index.has(nums[i])) {
      index.set(nums[i], [...index.get(nums[i]), i]);
    } else {
      index.set(nums[i], [i]);
    }
    let n = target - nums[i];
    if (n === nums[i]) {
      if (index.get(nums[i]).length >= 2) {
        result = [index.get(nums[i])[0], index.get(nums[i])[1]];
        break;
      }
    } else {
      if (index.has(n)) {
        result = [index.get(n)[0], index.get(nums[i])[0]];
        break;
      }
    }
  }
  return result;
};